File: data/lib/remote/paths.py
"""Path computation helpers for EFA V2 schema layout.

All paths are relative to the storage root (e.g. cache/remote/).
"""

from __future__ import annotations

from typing import TYPE_CHECKING


if TYPE_CHECKING:
    from pathlib import Path


def blob_path(root: Path, ident_hash: str, content_hash: str) -> Path:
    """assets/blobs/{2c}/{ident_hash}/{content_hash}"""
    return root / "assets" / "blobs" / ident_hash[:2] / ident_hash / content_hash


def blob_ident_dir(root: Path, ident_hash: str) -> Path:
    """assets/blobs/{2c}/{ident_hash}/"""
    return root / "assets" / "blobs" / ident_hash[:2] / ident_hash


def blob_shard_dir(root: Path, ident_hash: str) -> Path:
    """assets/blobs/{2c}/"""
    return root / "assets" / "blobs" / ident_hash[:2]


def resource_snapshot_dir(root: Path, snapshot_hash: str) -> Path:
    """assets/resources/{snapshot_hash}/"""
    return root / "assets" / "resources" / snapshot_hash


def release_snapshot_dir(root: Path, snapshot_hash: str) -> Path:
    """assets/releases/{snapshot_hash}/"""
    return root / "assets" / "releases" / snapshot_hash


def announcement_snapshot_dir(root: Path, snapshot_hash: str) -> Path:
    """assets/announcements/{snapshot_hash}/"""
    return root / "assets" / "announcements" / snapshot_hash


def generation_dir(root: Path, generation_hash: str) -> Path:
    """channels/refs/{generation_hash}/"""
    return root / "channels" / "refs" / generation_hash


def channel_head_dir(root: Path, channel: str) -> Path:
    """channels/heads/{channel}/"""
    return root / "channels" / "heads" / channel


def channel_registry_path(root: Path) -> Path:
    """channels/heads/channels.json"""
    return root / "channels" / "heads" / "channels.json"


def head_metadata_path(root: Path, channel: str) -> Path:
    """channels/heads/{channel}/metadata.json"""
    return channel_head_dir(root, channel) / "metadata.json"


def head_reflog_path(root: Path, channel: str) -> Path:
    """channels/heads/{channel}/reflog.pb2"""
    return channel_head_dir(root, channel) / "reflog.pb2"


# ---------------------------------------------------------------------------
# Temp / atomic write helpers
# ---------------------------------------------------------------------------


def temp_generation_dir(root: Path) -> Path:
    """channels/refs/tmp_gen/"""
    return root / "channels" / "refs" / "tmp_gen"


def temp_resource_snapshot_dir(root: Path) -> Path:
    """assets/resources/tmp_snapshot/"""
    return root / "assets" / "resources" / "tmp_snapshot"


def temp_release_snapshot_dir(root: Path) -> Path:
    """assets/releases/tmp_snapshot/"""
    return root / "assets" / "releases" / "tmp_snapshot"


def temp_announcement_snapshot_dir(root: Path) -> Path:
    """assets/announcements/tmp_snapshot/"""
    return root / "assets" / "announcements" / "tmp_snapshot"

File: lib/storage/repo/paths.dart
import "package:eve_fit_assistant/config/paths.dart";
import "package:path/path.dart" as p;

/// Path resolution for the EFA V2 unified storage schema.
///
/// All paths share the `resources/v2/` root. The layout follows
/// agent/schemav2/schema.md §2 (client-side).
class RepoPaths {
  const RepoPaths._();

  // ── Schema resources ────────────────────────────────────────────────────────

  static String get schemaResourcesPath => p.join(PathProvider.resourcesPath, "v2");

  static String get schemaVersionPath => p.join(schemaResourcesPath, "schema_version.json");

  // ── assets/ ─────────────────────────────────────────────────────────────────

  static String get assetsPath => p.join(schemaResourcesPath, "assets");

  /// Path to a blob in the shared content-addressed store.
  ///
  /// `identHash` is SHA-256 of the identifier URI (e.g. `resource://...`).
  /// `contentHash` is SHA-256 of the raw blob bytes.
  static String blobPath(String identHash, String contentHash) {
    if (identHash.length < 2) {
      throw ArgumentError.value(identHash, "identHash", "Must be at least 2 characters");
    }
    return p.join(assetsPath, "blobs", identHash.substring(0, 2), identHash, contentHash);
  }

  /// Directory containing all content versions for a single ident_hash.
  static String blobIdentDir(String identHash) {
    if (identHash.length < 2) {
      throw ArgumentError.value(identHash, "identHash", "Must be at least 2 characters");
    }
    return p.join(assetsPath, "blobs", identHash.substring(0, 2), identHash);
  }

  // ── assets/resources/ ───────────────────────────────────────────────────────

  /// Directory for a single resource snapshot.
  static String resourceSnapshotPath(String snapshotHash) =>
      p.join(assetsPath, "resources", snapshotHash);

  /// metadata.json for a resource snapshot.
  static String resourceSnapshotMetaPath(String snapshotHash) =>
      p.join(resourceSnapshotPath(snapshotHash), "metadata.json");

  /// resources.pb2 for a resource snapshot (ResourceIndex protobuf).
  static String resourceIndexPath(String snapshotHash) =>
      p.join(resourceSnapshotPath(snapshotHash), "resources.pb2");

  // ── announcements/ (client-side, per spec §2.3) ─────────────────────────────

  /// Root announcements directory.
  static String get announcementsRootPath => p.join(schemaResourcesPath, "announcements");

  static String announcementSnapshotPath(String snapshotHash) =>
      p.join(schemaResourcesPath, "announcements", snapshotHash);

  static String announcementSnapshotMetaPath(String snapshotHash) =>
      p.join(announcementSnapshotPath(snapshotHash), "metadata.json");

  static String announcementIndexPath(String snapshotHash) =>
      p.join(announcementSnapshotPath(snapshotHash), "announcements.pb2");

  // ── channels/ (client-side) ─────────────────────────────────────────────────

  static String get channelsPath => p.join(schemaResourcesPath, "channels");

  /// channels/channels.json — client channel registry.
  static String get channelRegistryPath => p.join(channelsPath, "channels.json");

  /// channels/{channel}/metadata.json — client channel head metadata.
  static String channelHeadMetaPath(String channelName) =>
      p.join(channelsPath, channelName, "metadata.json");

  /// channels/{channel}/server.pb2 — ServerIndex copied from generation.
  static String channelServerIndexPath(String channelName) =>
      p.join(channelsPath, channelName, "server.pb2");

  // ── checkouts/ (client-side) ────────────────────────────────────────────────

  static String get checkoutsPath => p.join(schemaResourcesPath, "checkouts");

  /// checkouts/checkouts.json — checkout registry.
  static String get checkoutRegistryPath => p.join(checkoutsPath, "checkouts.json");

  /// checkouts/{id}/metadata.json — individual checkout metadata.
  static String checkoutMetaPath(String checkoutId) =>
      p.join(checkoutsPath, checkoutId, "metadata.json");

  /// checkouts/{id}/reflog.pb2 — CheckoutReflog.
  static String checkoutReflogPath(String checkoutId) =>
      p.join(checkoutsPath, checkoutId, "reflog.pb2");
}
